SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUpstreamChecker.java

```diff
@@ -0,0 +1,81 @@
+package com.hubspot.singularity.scheduler;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.google.common.base.Optional;
+import com.hubspot.baragon.models.BaragonRequest;
+import com.hubspot.baragon.models.BaragonService;
+import com.hubspot.baragon.models.UpstreamInfo;
+import com.hubspot.singularity.LoadBalancerRequestType;
+import com.hubspot.singularity.LoadBalancerRequestType.LoadBalancerRequestId;
+import com.hubspot.singularity.SingularityDeploy;
+import com.hubspot.singularity.SingularityLoadBalancerUpdate;
+import com.hubspot.singularity.SingularityRequest;
+import com.hubspot.singularity.SingularityRequestWithState;
+import com.hubspot.singularity.SingularityTask;
+import com.hubspot.singularity.SingularityTaskId;
+import com.hubspot.singularity.data.DeployManager;
+import com.hubspot.singularity.data.RequestManager;
+import com.hubspot.singularity.data.TaskManager;
+import com.hubspot.singularity.hooks.LoadBalancerClientImpl;
+
+@Singleton
+public class SingularityUpstreamChecker {
+
+  private final LoadBalancerClientImpl lbClient;
+  private final TaskManager taskManager;
+  private final RequestManager requestManager;
+  private final DeployManager deployManager;
+
+  @Inject
+  public SingularityUpstreamChecker(LoadBalancerClientImpl lbClient, TaskManager taskManager, RequestManager requestManager, DeployManager deployManager) {
+    this.lbClient = lbClient;
+    this.taskManager = taskManager;
+    this.requestManager = requestManager;
+    this.deployManager = deployManager;
+  }
+
+  private List<SingularityTask> getActiveSingularityTasksForRequest (String requestId) {
+    List<SingularityTask> activeSingularityTasksForRequest = new ArrayList<>();
+    for (SingularityTaskId taskId: taskManager.getActiveTaskIdsForRequest(requestId)){
+      Optional<SingularityTask> maybeTask = taskManager.getTask(taskId);
+      if (maybeTask.isPresent()) {
+        activeSingularityTasksForRequest.add(maybeTask.get());
+      }
+    }
+    return activeSingularityTasksForRequest;
+  }
+
+  private Collection<UpstreamInfo> getUpstreamsFromActiveTasks(String requestId, Optional<String> loadBalancerUpstreamGroup) {
+    return lbClient.tasksToUpstreams(getActiveSingularityTasksForRequest(requestId), requestId, loadBalancerUpstreamGroup);
+  }
+
+  private SingularityLoadBalancerUpdate syncUpstreamsForService(SingularityRequest request, String requestId, SingularityDeploy deploy, Optional<String> loadBalancerUpstreamGroup) {
+    Collection<UpstreamInfo> upstreamsInBaragonForRequest = lbClient.getBaragonUpstreamsForRequest(requestId);
+    Collection<UpstreamInfo> upstreamsInSingularityForRequest = getUpstreamsFromActiveTasks(requestId, loadBalancerUpstreamGroup);
+    upstreamsInBaragonForRequest.removeAll(upstreamsInSingularityForRequest);
```

Comment:
we will only want to compare the `upstream` and `group` fields (maybe also `rackId` though not as important) of the `UpstreamInfo` object when determining if they are equal. The `requestId` field can be different and is an indicator of the unique baragon request id that originally added the upstream (not the Singularity request id). Right now we will not actually remove anything since the objects `equals` method also compares requestId